static/app/components/modals/inviteMembersModal/useInviteModal.tsx

```diff
@@ -0,0 +1,232 @@
+import {useIsMountedRef} from 'sentry/utils/useIsMountedRef';
+
+interface Props {
+  organization: Organization;
+  initialData?: Partial<InviteRow>[];
+  source?: string;
+}
+function defaultInvite(): InviteRow {
+  return {
+    emails: new Set<string>(),
+    teams: new Set<string>(),
+    role: 'member',
+  };
+}
+
+function useLogInviteModalOpened({
+  organization,
+  sessionId,
+  source,
+}: {
+  organization: Organization;
+  sessionId: string;
+  source: string | undefined;
+}) {
+  useEffect(() => {
+    trackAnalytics('invite_modal.opened', {
+      organization,
+      modal_session: sessionId,
+      can_invite: organization.access?.includes('member:write'),
+      source,
+    });
+  }, [organization, sessionId, source]);
+}
+
+export default function useInviteModal({organization, initialData, source}: Props) {
+  const api = useApi();
+  const willInvite = organization.access?.includes('member:write');
+  const isMounted = useIsMountedRef();
+
+  /**
+   * Used for analytics tracking of the modals usage.
+   */
+  const sessionId = useRef(uniqueId());
+  useLogInviteModalOpened({organization, sessionId: sessionId.current, source});
+
+  const memberResult = useApiQuery<Member>(
+    [`/organizations/${organization.slug}/members/me/`],
+    {
+      staleTime: 0,
+    }
+  );
+
+  const [state, setState] = useState(() => {
+    return {
+      pendingInvites: initialData
+        ? initialData.map(initial => ({
+            ...defaultInvite(),
+            ...initial,
+          }))
+        : [defaultInvite()],
+      inviteStatus: {},
+      complete: false,
+      sendingInvites: false,
+    };
+  });
+
+  const invites = useMemo(() => {
+    return state.pendingInvites.reduce<NormalizedInvite[]>(
+      (acc, row) =>
+        acc.concat(
+          Array.from(row.emails).map(email => ({email, teams: row.teams, role: row.role}))
+        ),
+      []
+    );
+  }, [state.pendingInvites]);
+
+  const reset = useCallback(() => {
+    setState({
+      pendingInvites: [defaultInvite()],
+      inviteStatus: {},
+      complete: false,
+      sendingInvites: false,
+    });
+    trackAnalytics('invite_modal.add_more', {
+      organization,
+      modal_session: sessionId.current,
+    });
+  }, [organization]);
+
+  const sendInvite = useCallback(
+    async (invite: NormalizedInvite) => {
+      const data = {
+        email: invite.email,
+        teams: [...invite.teams],
+        role: invite.role,
+      };
+
+      setState(prev => ({
+        ...prev,
+        inviteStatus: {
+          ...prev.inviteStatus,
+          [invite.email]: {sent: false},
+        },
+      }));
+
+      const endpoint = willInvite
+        ? `/organizations/${organization.slug}/members/`
+        : `/organizations/${organization.slug}/invite-requests/`;
+
+      try {
+        await api.requestPromise(endpoint, {method: 'POST', data});
+      } catch (err) {
+        if (!isMounted) {
+          return;
+        }
```

Comment:
I don't think this is needed since `useApi` cancels requests on unmount

Reply:
makes sense, because it was 'fine' before. I'm not trying to re-do all the logic here, just pour it into a functional component